### PR TITLE
SSCS-10075 Implement PUT partiesNotified Hearings Endpoint

### PR DESCRIPTION
**Pull request by dawidWalter**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/SSCS-10075

### Change description ###
Implement PUT partiesNotified Hearings Endpoint
Copied Dholaram's work from SSCS-10075-PartiesNotified-Put-Impl branch

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/contractTest/java/uk/gov/hmcts/reform/sscs/consumer/PartiesNotifiedPutPactConsumerTest.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        TimeUnit.SECONDS.sleep(2);
+    }
+
+    @Pact(provider = PROVIDER_NAME, consumer = CONSUMER_NAME)
+    public RequestResponsePact putPartiesNotifiedRequestForValidRequest(PactDslWithProvider builder) throws JsonProcessingException {
+        return builder.given(CONSUMER_NAME + " successfully putting a PartiesNotified request ")
+                .uponReceiving("Request to put PartiesNotified request")
+                .path(ContractTestDataProvider.PARTIES_NOTIFIED_PATH)
+                .method(HttpMethod.PUT.toString())
+                .query(FIELD_ID + "=" + VALID_CASE_ID + "&" + VERSION + "=" + VERSION_NUMBER)
```
Comment (TeaWithLucas) [inline src/contractTest/java/uk/gov/hmcts/reform/sscs/consumer/PartiesNotifiedPutPactConsumerTest.java:62]: Can you use String.format instead of concatenating strings?

Comment (TeaWithLucas) [reply src/contractTest/java/uk/gov/hmcts/reform/sscs/consumer/PartiesNotifiedPutPactConsumerTest.java:62]: Same with the ones below

File: src/contractTest/java/uk/gov/hmcts/reform/sscs/consumer/PartiesNotifiedPutPactConsumerTest.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+                .uponReceiving("Request to PUT PartiesNotified for bad PartiesNotified request")
+                .path(ContractTestDataProvider.PARTIES_NOTIFIED_PATH)
+                .method(HttpMethod.PUT.toString())
+                .query(FIELD_ID + "=" + VALID_CASE_ID + "&" + VERSION + "=" + VERSION_NUMBER)
+                .body(ContractTestDataProvider.toJsonString(ContractTestDataProvider.generateInvalidPartiesPutRequest()))
+                .headers(ContractTestDataProvider.authorisedHeaders)
+                .willRespondWith()
+                .status(HttpStatus.BAD_REQUEST.value())
+                .body(new PactDslJsonBody()
+                        .stringType(ContractTestDataProvider.FIELD_MESSAGE, ContractTestDataProvider.MSG_400_HEARING)
```
Comment (TeaWithLucas) [inline src/contractTest/java/uk/gov/hmcts/reform/sscs/consumer/PartiesNotifiedPutPactConsumerTest.java:83]: can you import the static instead of using the ContractTestDataProvider qualifier

File: src/main/java/uk/gov/hmcts/reform/sscs/service/HearingsService.java [hunk 3]
```diff
@@ -199,4 +204,19 @@ private HearingWrapper createWrapper(HearingRequest hearingRequest) throws GetCa
                 .sscsServiceCode(sscsServiceCode)
                 .build();
     }
+
+    private void sendPartiesNotifiedUpdateRequest(HearingWrapper wrapper) {
```
Comment (TeaWithLucas) [inline src/main/java/uk/gov/hmcts/reform/sscs/service/HearingsService.java:208]: Can you put this method with the other similar methods above

Comment (TeaWithLucas) [inline src/main/java/uk/gov/hmcts/reform/sscs/service/HearingsService.java:208]: This should return a Response, such as PartiesNotifiedHearingResponse

File: src/main/java/uk/gov/hmcts/reform/sscs/service/HearingsService.java [hunk 4]
```diff
@@ -199,4 +204,19 @@ private HearingWrapper createWrapper(HearingRequest hearingRequest) throws GetCa
...
+                getHearingId(wrapper),
+                getVersionNumber(wrapper),
+                buildUpdatePartiesNotifiedPayload(wrapper)
+
+        );
+    }
+
+    private IdamTokens receiveIdamToken() {
+        return idamService.getIdamTokens();
+    }
```
Comment (TeaWithLucas) [inline src/main/java/uk/gov/hmcts/reform/sscs/service/HearingsService.java:221]: This method is not needed, just use idamService.getIdamTokens(). ... instead

File: src/main/java/uk/gov/hmcts/reform/sscs/service/HmcHearingPartiesNotifiedApi.java [hunk 5]
```diff
@@ -3,25 +3,38 @@
...
+import org.springframework.web.bind.annotation.RequestBody;
 import org.springframework.web.bind.annotation.RequestHeader;
 import org.springframework.web.bind.annotation.RequestParam;
 import uk.gov.hmcts.reform.sscs.config.FeignClientConfig;
-import uk.gov.hmcts.reform.sscs.model.single.hearing.PartiesNotified;
+import uk.gov.hmcts.reform.sscs.model.partiesnotified.PartiesNotified;
 
 import static org.springframework.http.HttpHeaders.AUTHORIZATION;
 
+@SuppressWarnings({"PMD.UseObjectForClearerAPI"})
```
Comment (TeaWithLucas) [inline src/main/java/uk/gov/hmcts/reform/sscs/service/HmcHearingPartiesNotifiedApi.java:15]: Please implement this rather than supressing it

File: src/test/java/uk/gov/hmcts/reform/sscs/service/HearingsServiceTest.java [hunk 6]
```diff
@@ -118,6 +120,24 @@ void processHearingRequest(HearingState state) throws GetCaseException, InvalidI
                 .isThrownBy(() -> hearingsService.processHearingRequest(request));
     }
 
+    @DisplayName("When wrapper with a PARTY_NOTIFIED Hearing State is given partyNotified should run without error")
+    @ParameterizedTest
+    @CsvSource(value = {
+        "PARTY_NOTIFIED"
+    }, nullValues = {"null"})
```
Comment (TeaWithLucas) [inline src/test/java/uk/gov/hmcts/reform/sscs/service/HearingsServiceTest.java:127]: ParameterizedTest is not needed for one value

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/PartiesNotifiedMappingTest.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        PartiesNotified partiesNotified = PartiesNotifiedMapping.buildUpdatePartiesNotifiedPayload(wrapper);
+
+        assertNotNull(partiesNotified.getRequestVersion());
+    }
+
+
+    @DisplayName("getVersionNumber Test")
+    @Test
+    void getVersionNumber() {
+        // TODO Finish Test when method done
```
Comment (TeaWithLucas) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/PartiesNotifiedMappingTest.java:40]: Is this test finalised now?

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/PartiesNotifiedMappingTest.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+class PartiesNotifiedMappingTest extends HearingsMappingBase {
+    public static final String VERSION_NUMBER = "1";
+
+    @DisplayName("When a valid hearing wrapper is given buildUpdatePartiesNotifiedPayload returns the correct Hearing Request")
+    @Test
+    void buildUpdatePartiesNotifiedPayload() {
+
+        // TODO Finish Test when method done
```
Comment (TeaWithLucas) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/PartiesNotifiedMappingTest.java:20]: Is this test finalised now?

**Review by TeaWithLucas — CHANGES_REQUESTED** [review]

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/service/HearingsServiceHelperTest.java [hunk 9]
```diff
@@ -139,4 +139,6 @@ private HearingWrapper activeHearingIdFixture(final Long hearingId) {
                     .build())
             .build();
     }
+
```
Comment (suetlaiHMCTS) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/service/HearingsServiceHelperTest.java:142]: May I know is it able to remove the space line?

File: src/main/java/uk/gov/hmcts/reform/sscs/service/HearingsService.java (hunk 4 above)
Comment (dattipoeHMCTS) [reply src/main/java/uk/gov/hmcts/reform/sscs/service/HearingsService.java:221]: Does it? I'm curious about this...I need to get better at LoD

File: src/main/java/uk/gov/hmcts/reform/sscs/model/partiesnotified/PartiesNotified.java [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+import lombok.NoArgsConstructor;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@Builder(toBuilder = true)
+public class PartiesNotified {
+    private String requestVersion;
+    private JsonNode serviceData;
```
[the remark this answers is not in the document]
Comment (augustio) [reply src/main/java/uk/gov/hmcts/reform/sscs/model/partiesnotified/PartiesNotified.java:17]: Any idea what the fields would be? I think the idea idea here is for it to be generic and be able to take any Json object?

File: src/main/java/uk/gov/hmcts/reform/sscs/service/HearingsService.java (hunk 3 above)
Comment (augustio) [reply src/main/java/uk/gov/hmcts/reform/sscs/service/HearingsService.java]: This calls the client method that performs a PUT operation, updating an existing resource. Does it really need to return something? This is also not in the spec.

**Review by nickwest1 — DISMISSED** [review]

File: src/contractTest/java/uk/gov/hmcts/reform/sscs/consumer/PartiesNotifiedPutPactConsumerTest.java [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+        TimeUnit.SECONDS.sleep(2);
+    }
+
+    @Pact(provider = PROVIDER_NAME, consumer = CONSUMER_NAME)
+    public RequestResponsePact putPartiesNotifiedRequestForValidRequest(PactDslWithProvider builder) throws JsonProcessingException {
+        return builder.given(CONSUMER_NAME + " successfully putting a PartiesNotified request ")
+                .uponReceiving("Request to put PartiesNotified request")
+                .path(PARTIES_NOTIFIED_PATH)
+                .method(HttpMethod.PUT.toString())
+                .query(String.format("%s=%s&%s=%s", FIELD_ID, VALID_CASE_ID, VERSION, VERSION_NUMBER))
```
Comment (TeaWithLucas) [inline src/contractTest/java/uk/gov/hmcts/reform/sscs/consumer/PartiesNotifiedPutPactConsumerTest.java:77]: Can you refactor the template "%s=%s&%s=%s" as a constant?

File: src/contractTest/java/uk/gov/hmcts/reform/sscs/consumer/PartiesNotifiedPutPactConsumerTest.java [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+                .body(toJsonString(generateInvalidPartiesPutRequest()))
+                .headers(authorisedHeaders)
+                .willRespondWith()
+                .status(HttpStatus.BAD_REQUEST.value())
+                .body(new PactDslJsonBody()
+                        .stringType(FIELD_MESSAGE, MSG_400_HEARING)
+                        .stringValue(FIELD_STATUS, HttpStatus.BAD_REQUEST.value()
+                                + " " + HttpStatus.BAD_REQUEST.getReasonPhrase())
```
Comment (TeaWithLucas) [inline src/contractTest/java/uk/gov/hmcts/reform/sscs/consumer/PartiesNotifiedPutPactConsumerTest.java:100]: can you string.format these in this class and use a constant for the template?

**Review by TeaWithLucas — APPROVED** [review]

File: src/contractTest/java/uk/gov/hmcts/reform/sscs/consumer/PartiesNotifiedPutPactConsumerTest.java [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+
+
+@ExtendWith(PactConsumerTestExt.class)
+@EnableFeignClients(basePackages = {"uk.gov.hmcts.reform.sscs.service"})
+@ActiveProfiles("contract")
+@SpringBootTest
+@PactTestFor(port = "10000")
+@PactFolder("pacts")
+class PartiesNotifiedPutPactConsumerTest extends BasePactTest {
+    public static final String FIELD_STATUS_FORMAT = "%d %s";
```
Comment (nickwest1) [inline src/contractTest/java/uk/gov/hmcts/reform/sscs/consumer/PartiesNotifiedPutPactConsumerTest.java:58]: Why public?

File: src/contractTest/java/uk/gov/hmcts/reform/sscs/consumer/PartiesNotifiedPutPactConsumerTest.java [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+    private static final String VERSION = "version";
+    private static final Long VERSION_NUMBER = 123L;
+    private static final String FORBIDDEN_CASE_ID = "456";
+    private static final String NOT_FOUND_CASE_ID = "789";
+    public static final String QUERY_FORMAT = "%s=%s&%s=%s";
```
Comment (nickwest1) [inline src/contractTest/java/uk/gov/hmcts/reform/sscs/consumer/PartiesNotifiedPutPactConsumerTest.java:63]: Why public?

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/PartiesNotifiedMappingTest.java [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+import uk.gov.hmcts.reform.sscs.ccd.domain.SchedulingAndListingFields;
+import uk.gov.hmcts.reform.sscs.ccd.domain.SscsCaseData;
+import uk.gov.hmcts.reform.sscs.model.HearingWrapper;
+import uk.gov.hmcts.reform.sscs.model.partiesnotified.PartiesNotified;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+class PartiesNotifiedMappingTest extends HearingsMappingBase {
+    public static final Long VERSION_NUMBER = 1L;
```
Comment (nickwest1) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/PartiesNotifiedMappingTest.java:14]: Why public?

**Review by nickwest1 — APPROVED** [review]
Just a few trivial comments that happy if you apply or not. Worth at least considering for future dev though ;)

File: src/contractTest/java/uk/gov/hmcts/reform/sscs/consumer/PartiesNotifiedPutPactConsumerTest.java (hunk 13 above)
Comment (augustio) [reply src/contractTest/java/uk/gov/hmcts/reform/sscs/consumer/PartiesNotifiedPutPactConsumerTest.java:58]: I guess it is better made private. Done.

File: src/contractTest/java/uk/gov/hmcts/reform/sscs/consumer/PartiesNotifiedPutPactConsumerTest.java (hunk 14 above)
Comment (augustio) [reply src/contractTest/java/uk/gov/hmcts/reform/sscs/consumer/PartiesNotifiedPutPactConsumerTest.java:63]: I guess it is better made private. Done.

**Review by nickwest1 — DISMISSED** [review]

File: src/main/java/uk/gov/hmcts/reform/sscs/service/HearingsService.java [hunk 16]
```diff
@@ -118,7 +122,7 @@ private void cancelHearing(HearingWrapper wrapper) {
     }
 
     private void partyNotified(HearingWrapper wrapper) {
-        // TODO SSCS-10075 - implement mapping for the event when a party has been notified, might not be needed
+        sendPartiesNotifiedUpdateRequest(wrapper);
```
Comment (MohammedHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/service/HearingsService.java:125]: Why is this nested why isn't this method just in one? I know the rule is one method should have one purpose but the only purpose this has is to call another method that has one purpose?

File: src/main/java/uk/gov/hmcts/reform/sscs/model/partiesnotified/PartiesNotified.java [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@Builder(toBuilder = true)
+public class PartiesNotified {
```
Comment (MohammedHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/model/partiesnotified/PartiesNotified.java:14]: We already have a PartiesNotified class why was this added and the other not removed or altered?

**Review by MohammedHMCTS — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/sscs/service/HearingsService.java [hunk 18]
```diff
@@ -166,6 +169,16 @@ public HearingResponse sendCancelHearingRequest(HearingWrapper wrapper) {
         );
     }
 
+    private void sendPartiesNotifiedUpdateRequest(HearingWrapper wrapper) {
+        hmcHearingPartiesNotifiedApi.updatePartiesNotifiedHearingRequest(
```
Comment (TeaWithLucas) [inline src/main/java/uk/gov/hmcts/reform/sscs/service/HearingsService.java:173]: Don't we need to get a response here to make sure the request has gone through okay?

File: src/main/java/uk/gov/hmcts/reform/sscs/service/HearingsService.java (hunk 16 above)
Comment (TeaWithLucas) [reply src/main/java/uk/gov/hmcts/reform/sscs/service/HearingsService.java:128]: There should be a response here and related processing

File: src/main/java/uk/gov/hmcts/reform/sscs/model/partiesnotified/PartiesNotified.java (hunk 17 above)
Comment (TeaWithLucas) [reply src/main/java/uk/gov/hmcts/reform/sscs/model/partiesnotified/PartiesNotified.java:14]: Where?

**Review by TeaWithLucas — APPROVED** [review]
